sympy/matrices/tests/test_random.py

```diff
@@ -0,0 +1,535 @@
+
+
+def test_diagonalizable():
+    for d in TEST_DIMS:
+        m = diagonalizable(d, spec=(2,))
+        assert m.eigenvals(multiple=True) == [2] * d
+        for r in range(1, d):
+            s = (0,) * (d - r) + (2,) * r
+            m = diagonalizable(d, spec=s)
+            assert m.rank() == r
+            for ev in m.eigenvals(multiple=True):
+                assert ev in (0, 2), m.eigenvals(multiple=True)
+
+
+def test_trigonalizable():
+    for d in TEST_DIMS:
+        m = trigonalizable(d, spec=(2,))
+        assert m.eigenvals(multiple=True) == [2] * d
+        for r in range(1, d):
+            s = (0, None) * (d - r) + (2,) * r
+            m = trigonalizable(d, spec=s)
+            assert m.rank() == r, (m.rank(), r)
+            for ev in m.eigenvals(multiple=True):
+                assert ev in (0, 2), m.eigenvals(multiple=True)
+
+
+# === conjugate matrices, i.e. defined by similarity to a normal from ==
+
+
+def test_idempotent():
+    for d in TEST_DIMS:
+        for r in range(1, d):
+            m = idempotent(d, r)
+            assert m * m == m
+            assert m.rank() == r
+
+
+def test_nilpotent():
+    for d in TEST_DIMS:
+        for r in range(1, d):
+            m = nilpotent(d, r)
+            assert m.rank() == r
+            assert _is_zeros(m ** d)
+
+
+# === matrices conjugate by isometries ==
+
+
+def test_orthogonal():
+    for d in TEST_DIMS:
+        i = orthogonal(d, (1,), length=0)
+        assert _is_eye(i), repr(i)
+
+        m = orthogonal(d).evalf()
+        assert _is_eye(m.T * m, TEST_PRECISION)
+        assert abs(m.evalf().det()) - 1 < TEST_PRECISION
+
+        m = orthogonal(d, spec=_random._rotation_scalar_set).evalf()
+        assert _is_eye(m.T * m, TEST_PRECISION)
+        assert abs(m.evalf().det()) - 1 < TEST_PRECISION
+
+        m = orthogonal(d, spec=(cos(phi),))
+        assert _is_eye(m.T * m)
+
+
+def test_unitary():
+    for d in list(TEST_DIMS)[1::3]:
+        i = unitary(d, (1,), length=0)
+        assert _is_eye(i), repr(i)
+
+        m = unitary(d)
+        assert _is_eye(m.H * m, TEST_EPSILON), repr(simplify(m.H * m))
+        assert _is_eye(m.H * m), repr((m.H * m).evalf())
+        assert abs(abs(m.det()) - 1) < TEST_EPSILON
+
+        z = complex(1, 2)
+        m = unitary(d, spec=(z / abs(z),), length=d)
+        assert _is_eye(m.H * m, TEST_EPSILON), repr(simplify(m.H * m))
+        assert abs(m.evalf().det()) - 1 < TEST_PRECISION
+
+        z = 2 + 3 * I
+        m = unitary(d, spec=(z / abs(z),), length=d)
+        assert _is_eye(m.H * m), repr(simplify(m.H * m))
+
+        m = unitary(d, spec=(sqrt(2) / 2,), length=d)
+        assert _is_eye(m.H * m), repr(simplify(m.H * m))
+
+
+def test_normal():
+    for d in list(TEST_DIMS)[1::3]:
+        spec = (2, 3, 4)
+        m = normal(d, spec, length=d)
+        assert all(x.is_real for x in simplify(m)), repr(simplify(m))
+        assert _is_zeros(m.T * m - m * m.T), \
+            repr(simplify(m.T * m - m * m.T))
```

Comment:
With random tests like this we need to make sure that when the test fails the error message gives a way to reproduce the failure (e.g. it prints the seed or something). Otherwise if you run the tests once and it fails then you have no way to reproduce that failure afterwards.

Reply:
seed is set once for all tests at the top of the test file
https://github.com/sympy/sympy/pull/22099/files/3375a9ded74f4a3350e74ea1a042c199fc18b138#r757045756
so any failure could be reproduced. 

In case of a failure, one can add some prints as done here already
https://github.com/sympy/sympy/pull/22099/files/3375a9ded74f4a3350e74ea1a042c199fc18b138#r757047379

Since there are plenty of tests (more than 600 lines of code) I don't think it would be efficient to add prints to all of them now. 

But if there are special reason to do this only for test of `normal` I think it would be feasible.